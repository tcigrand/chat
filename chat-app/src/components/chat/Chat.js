import React, { Component } from "react";
import { Box, Button, TextField } from "@material-ui/core";
import SendIcon from "@material-ui/icons/Send";

import {
  sendMessage,
  subscribeToMessagesForUserAndRecipient,
  removeAllListeners
} from "./../../services/chat-sockets";

class Chat extends Component {
  constructor(props) {
    super(props);
    this.state = { messages: [], messageToSend: "" };
  }

  componentWillReceiveProps(props) {
    removeAllListeners();
    this.setState({ messages: [], messageToSend: "" });

    subscribeToMessagesForUserAndRecipient(
      props.currentUser,
      props.recipient,
      data => {
        this.setState(prevState => {
          const newState = prevState;
          newState.messages.push(data);

          return newState;
        });
      }
    );
  }

  sendMessage = () => {
    if (this.state.messageToSend) {
      const message = {
        message: this.state.messageToSend,
        from: this.props.currentUser,
        to: this.props.recipient
      };
      sendMessage(message);
      this.setState(prevState => {
        const newState = prevState;
        newState.messages.push(message);
        newState.messageToSend = "";
        return newState;
      });
    }
  };

  setMessageToSend = $event => {
    this.setState({
      messageToSend: $event.target.value
    });
  };

  componentWillUnmount() {
    removeAllListeners();
  }

  renderMessages() {
    return this.state.messages.map((data, i) => {
      return (
        <div key={i}>
          <b>{data.from}: </b>
          {data.message}
        </div>
      );
    });
  }

  render() {
    if (this.props.recipient) {
      return (
        <Box
          display="flex"
          flexDirection="column"
          bgcolor="lightgray"
          color="text.primary"
          px={2}
          height={1}
        >
          <h3>Your chat with {this.props.recipient}</h3>
          <Box
            display="flex"
            flexDirection="column"
            bgcolor="white"
            px={2}
            height={1}
          >
            <Box component="div" overflow="scroll" maxHeight={"50%"}>
              {this.renderMessages()}
            </Box>
            <TextField
              id="outlined-multiline-static"
              label="Type here"
              multiline
              rows="4"
              margin="normal"
              variant="outlined"
              value={this.state.messageToSend}
              onChange={this.setMessageToSend}
            />
            <Button
              variant="contained"
              color="primary"
              onClick={this.sendMessage}
            >
              Send
              <SendIcon />
            </Button>
          </Box>
        </Box>
      );
    }
    return null;
  }
}

export default Chat;
