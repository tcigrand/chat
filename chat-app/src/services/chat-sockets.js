import openSocket from "socket.io-client";
const socket = openSocket("http://localhost:8000");

function listUsers(cb) {
  socket.emit("listUsers");
  socket.on("users", users => cb(users));
}

function userLogin(user) {
  socket.emit("userLogin", { user });
}

function sendMessage(data) {
  socket.emit("messageSent", data);
}

function subscribeToMessagesForUserAndRecipient(currentUser, from, cb) {
  socket.on("messageReceived", data => {
    if (currentUser === data.to && from === data.from) {
      cb(data);
    }
  });
}

function removeAllListeners() {
  socket.removeAllListeners();
}

export {
  removeAllListeners,
  subscribeToMessagesForUserAndRecipient,
  sendMessage,
  listUsers,
  userLogin
};
