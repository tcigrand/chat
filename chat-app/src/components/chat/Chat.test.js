import React from "react";
import { shallow } from "enzyme";
import Chat from "./Chat";

import * as socketCalls from "./../../services/chat-sockets";

let wrapper;

beforeEach(() => {
  wrapper = shallow(<Chat recipient="Bob" currentUser="Linda" />);
});

describe("Chat", () => {
  it("renders without crashing", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("renders welcome message", () => {
    const yourChat = <h3>Your chat with Bob</h3>;
    expect(wrapper.contains(yourChat)).toEqual(true);
  });

  it("sends message with correct format", () => {
    wrapper.instance().setState({ messageToSend: "Hi" });
    const sendMessageSpy = spyOn(socketCalls, "sendMessage");
    wrapper.instance().sendMessage();
    expect(sendMessageSpy).toHaveBeenCalledWith({
      from: "Linda",
      message: "Hi",
      to: "Bob"
    });
    expect(wrapper.instance().state.messageToSend).toBe("");
  });

  it("subscribes to messages for current user + recipient", () => {
    const subscribeToMessagesForUserAndRecipientSpy = spyOn(
      socketCalls,
      "subscribeToMessagesForUserAndRecipient"
    );
    const removeAllListenersSpy = spyOn(socketCalls, "removeAllListeners");
    wrapper.setProps({ recipient: "Tim", currentUser: "Tim2" });
    expect(removeAllListenersSpy).toHaveBeenCalled();
    expect(subscribeToMessagesForUserAndRecipientSpy).toHaveBeenCalledWith(
      "Tim2",
      "Tim",
      expect.any(Function)
    );
  });

  it("changing props resets state", () => {
    wrapper
      .instance()
      .setState({ messageToSend: "Hi Again", messages: ["test"] });
    expect(wrapper.instance().state).toEqual({
      messageToSend: "Hi Again",
      messages: ["test"]
    });
    wrapper.setProps({ recipient: "Tim", currentUser: "Tim2" });
    expect(wrapper.instance().state).toEqual({
      messageToSend: "",
      messages: []
    });
  });

  it("renders messages correctly", () => {
    wrapper.instance().setState({ messageToSend: "Hi" });
    wrapper.instance().sendMessage();

    const chat = (
      <div>
        <b>Linda: </b>
        Hi
      </div>
    );
    expect(wrapper.contains(chat)).toEqual(true);
  });
});
