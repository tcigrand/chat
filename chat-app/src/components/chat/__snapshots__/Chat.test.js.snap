// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat renders without crashing 1`] = `ShallowWrapper {}`;

exports[`renders without crashing 1`] = `ShallowWrapper {}`;
